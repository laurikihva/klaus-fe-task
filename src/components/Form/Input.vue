<template>
  <div class="flex flex-col gap-2 text-sm">
    <label v-if="label" :for="id">{{ label }}</label>
    <div class="relative text-gray-600">
      <div v-if="icon" class="absolute top-1/2 -translate-y-1/2 left-2 pointer-events-none">
        <Icon :name="icon" />
      </div>
      <input :id v-model="model" :name :class="computedInputClasses" v-bind="attrs" />
    </div>
  </div>
</template>

<script lang="ts" setup>
import { toRefs, computed, useAttrs } from "vue";

import Icon, { TIconName } from "@/components/Icon.vue";

export interface IInputProps {
  id: string;
  label?: string;
  name?: string;
  icon?: TIconName;
}

const props = defineProps<IInputProps>();
const { icon } = toRefs(props);

defineOptions({
  inheritAttrs: false,
});

const model = defineModel<string | number>({ required: true });

const attrs = useAttrs();

const computedInputClasses = computed(() => {
  const base =
    "w-full min-h-[40px] pr-2 py-2 placeholder:text-gray-100 enabled:focus:outline-blue-400 rounded bg-white disabled:opacity-50 border border-gray-400";
  const withIcon = icon.value ? "pl-8" : "pl-2";

  return [base, withIcon];
});
</script>
