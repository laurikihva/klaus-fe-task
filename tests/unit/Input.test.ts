import { mount } from "@vue/test-utils";

import Input from "@/components/Form/Input.vue";

const wrapper = mount(Input, {
  props: {
    id: "my-input",
    modelValue: "",
    label: "Input label",
    name: "my-input",
    icon: "search",
  },
});

describe("Loads and displays Input component", () => {
  it("renders correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("contains svg, label and base classes", () => {
    expect(wrapper.find("svg").exists()).toBeTruthy();
    expect(wrapper.text()).toContain("Input label");

    expect(wrapper.classes("flex")).toBe(true);
    expect(wrapper.classes("flex-col")).toBe(true);
  });
});
