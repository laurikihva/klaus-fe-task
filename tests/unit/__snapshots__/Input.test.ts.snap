// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loads and displays Input component > renders correctly 1`] = `
"<div class="flex flex-col gap-2 text-sm"><label for="my-input">Input label</label>
  <div class="relative text-gray-600">
    <div class="absolute top-1/2 -translate-y-1/2 left-2 pointer-events-none"><svg xmlns="http://www.w3.org/2000/svg" fill="currentColor" viewBox="0 0 16 16" class="w-[14px] h-[14px]">
        <path d="m15.805 14.863-3.98-3.98a6.677 6.677 0 1 0-.942.943l3.979 3.98a.667.667 0 0 0 .943-.943M6.667 12A5.334 5.334 0 1 1 12 6.668 5.34 5.34 0 0 1 6.667 12"></path>
      </svg></div><input id="my-input" name="my-input" class="w-full min-h-[40px] pr-2 py-2 placeholder:text-gray-100 enabled:focus:outline-blue-400 rounded bg-white disabled:opacity-50 border border-gray-400 pl-8">
  </div>
</div>"
`;
